File: capitulo1/pesquisa/pesquisa_binaria.c
#include <stdio.h>

void pesquisa_binaria(int item, int vetor[], int len_vector);

#define TAM 10

int main(){

    int vetor[TAM] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
    int item;
    
    printf("Informe um número: ");
    scanf("%d", &item);

    pesquisa_binaria(item, vetor, TAM);

    return 0;
}

void pesquisa_binaria(int item, int vetor[], int len_vetor){
    int baixo = 0;
    int alto = len_vetor - 1;

    int meio;

    while(baixo <= alto){

        meio = (int) ((baixo + alto) / 2);

        if(vetor[meio] == item){
            printf("Item (%d) encontrado no índice %d\n", item, meio);
            return;
        }

        else if(item > vetor[meio]){
            baixo = meio + 1;
        }

        else{
            alto = meio - 1;
        }

    }
    printf("%d não está na lista!\n", item);

}

File: capitulo1/pesquisa/pesquisa_binaria.py
def pesquisa_binaria(lista: list, elemento: int) -> int:

    tam_lista = len(lista)

    baixo = 0 # Primeiro elemento da lista

    alto = tam_lista - 1 # Último elemento da lista

    while(baixo <= alto):

        chute = (baixo + alto) // 2 # meio da lista

        if lista[chute] == elemento:

            return chute
            
        elif elemento > lista[chute]:

            baixo = chute + 1

        else:

            alto = chute - 1

    return None


a = int(input("Informe o início do intervalo: "))
b = int(input("Informe o fim do intervalo: "))

lista = [i for i in range(a, b)]
elemento = int(input("Informe o elemento que deseja encontrar o índice: "))

result = pesquisa_binaria(lista=lista, elemento=elemento)

if result:

    print(f"{elemento} encontrado no índice: {result}")

else:

    print(f"{elemento} não faz parte da lista!")

